File: javascript/grind-grokking-coding-interview/remove-duplicates.js
/**
 * Given an array of sorted numbers, remove all duplicates from it. You should not use any extra space; after removing the duplicates in-place return the new length of the array.
 */

function removeDuplicates(arr) {
  let nextNonDuplicate = 1; // this is the index of the slot that we're going to put the next non-duplicated in

  let next = 1; // this is the traveler. It's initialized 1 because the first element in the array always be counted.
  while (next < arr.length) {
    if (arr[nextNonDuplicate - 1] !== arr[next]) {
      arr[nextNonDuplicate] = arr[next];
      nextNonDuplicate += 1;
    }
    next += 1;
  }

  return nextNonDuplicate;
}

console.log(removeDuplicates([2, 3, 3, 3, 6, 9, 9]));
console.log(removeDuplicates([2, 2, 2, 11]));

File: problems/java/29012024-two-pointers-removeduplicates/RemoveDuplicates.java
/**
 * Constraint: not use any extra space, and input array is already sorted
 */
public class RemoveDuplicates {
  /**
   * Time complexity: O(n)
   * Space complexity: O(1)
   */
  public static int remove(int[] arr) {
    int nextNonDuplicate = 1;

    for (int i = 1; i < arr.length; i++) {
      if (arr[i - 1] != arr[i]) {
        arr[nextNonDuplicate] = arr[i];
        nextNonDuplicate++;
      }
    }

    return nextNonDuplicate;
  }

  public static void main(String[] args) {
    System.out.println(remove(new int[] { 2, 3, 3, 3, 6, 9, 9 }));
    System.out.println(remove(new int[] { 2, 2, 2, 11 }));
  }
}
